[WIP] X509: Refactor X509_PUBKEY processing to include provider side keys

levitte opened the pull request:
When a SubjectPublicKeyInfo (SPKI) is decoded into an X509_PUBKEY
structure, the corresponding EVP_PKEY is automatically added as well.
This used to only support our built-in keytypes, and only in legacy
form.

This is now refactored to try to decode the whole SPKI into an
EVP_PKEY using OSSL_DECODER, and cache that in the X509_PUBKEY very
early.  Only when that fails, the existing decoder callback will try
to create a legacy EVP_PKEY.

For purposes where it's not desirable to do the provider side SPKI
decoding, there is now a function d2i_PUBKEY_legacy().  This is used
by all the type specific d2i_<TYPE>_PUBKEY(), as well as in out
provider DER->key decoder, to avoid running into recursion problems.

Fixes #13893

-----

This includes refactoring the DER-\>key decoder to be a bit more flexible with how it uses keytype specific d2i functions and adds the possibility to use d2i_*TYPE*_PUBKEY.

This also includes the addition of SM2 encoders and decoders...

levitte committed:
Fix some odd names in our provider source code

ecossl_dh_keyexch_functions     -> ossl_ecdh_keyexch_functions
ecossl_dsa_signature_functions  -> ossl_ecdsa_signature_functions
sm2_asym_cipher_functions       -> ossl_sm2_asym_cipher_functions
sm2_keymgmt_functions           -> ossl_sm2_keymgmt_functions
sm2_signature_functions         -> ossl_sm2_signature_functions

levitte committed:
PROV: Add SM2 encoders and decoders, as well as support functionality

The EC KEYMGMT implementation handled SM2 as well, except what's
needed to support decoding: loading functions for both EC and SM2 that
checks for the presence or absence of the SM2 curve the same way as
the EC / SM2 import functions.

levitte committed:
PROV: Refactor DER->key decoder

The decoding of DER into keys with keytype specific routines depended
entirely on the absence of the generic algo specific C type from
EVP_PKEYs.  That is not necessary, and may even prove to be a bit of a
disadvantage, depending on what libcrypto has to offer in terms of
type specific d2i functionality for different kinds of input
structures.

To remedy, we try with all available type specific functions first,
and only turn to the general d2i functions (those that return an
EVP_PKEY) as a last resort.

levitte committed:
PROV: Add type specific SubjectPublicKeyInfo decoding to the DER->key decoders

This makes it possible to use d2i_<TYPE>_PUBKEY instead of the generic
d2i_PUBKEY()

levitte committed:
X509: Refactor X509_PUBKEY processing to include provider side keys

When a SubjectPublicKeyInfo (SPKI) is decoded into an X509_PUBKEY
structure, the corresponding EVP_PKEY is automatically added as well.
This used to only support our built-in keytypes, and only in legacy
form.

This is now refactored to try to decode the whole SPKI into an
EVP_PKEY using OSSL_DECODER, and cache that in the X509_PUBKEY very
early.  Only when that fails, the existing decoder callback will try
to create a legacy EVP_PKEY.

For purposes where it's not desirable to do the provider side SPKI
decoding, there is now a function d2i_PUBKEY_legacy().  This is used
by all the type specific d2i_<TYPE>_PUBKEY(), as well as in out
provider DER->key decoder, to avoid running into recursion problems.

Fixes #13893

levitte commented:
Something to be noted is the inherent problem with the early caching of an EVP_PKEY in the X509_PUBKEY, when someone else wants an easy way to decode the DER, as part of a decoder implementation for a keytype we know nothing about.  It becomes a recursion problem.

The solution I could come up with was to have a `d2i_PUBKEY_legacy()`, which is currently internal.  However, that doesn't help anyone making an external provider for a keytype that's unknown to us, and that merely wants to decode the SubjectPublicKeyInfo into its parts, but no further processing (i.e. no automatic caching of an EVP_PKEY).

levitte commented:
A slightly different solution could have been to let `x509_public_decode()` handle the decoding via OSSL_DECODER.  The only issue there is that keytypes like RSA-PSS would be a bit problematic, as its parameters aren't included in the pubkey BIT STRING itself, but rather in the AlgorithmID (X509_ALGOR) that sits beside it.

This could be resolved by decoding the X509_ALGOR `parameters` field and the X509_PUBKEY `pubkey` fields separately, which is possible with the help of d2i_*TYPE*PublicKey() and d2i_*TYPE*params(), assign what we get to EVP_PKEYs, and use EVP_PKEY_copy_parameters() to copy the parameters extracted from the X509_ALGOR onto the EVP_PKEY that contains the key itself...  the only issue there being we don't have  d2i_*TYPE*PublicKey() and d2i_*TYPE*params() for all our legacy keytypes.

mattcaswell commented:
Hmm, this solution doesn't seem to work for me. If I run this code:

````
        x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
        assert(x509 != NULL);
        pkey = X509_get_pubkey(x509);

        assert(pkey != NULL);
        assert(pkey->keymgmt != NULL);
````

I get:
````
readx509-2: readx509-2.c:24: main: Assertion `pkey->keymgmt != NULL' failed.
Aborted (core dumped)
````

levitte commented:
Yeah, something is off. See the CI failures

levitte commented:
What key type was there in that certificate?  I just found a problem where RSA and RSA-PSS gets mixed up...

mattcaswell commented:
RSA

levitte committed:
PROV: Add RSA-PSS specific OSSL_FUNC_KEYMGMT_LOAD function

The OSSL_FUNC_KEYMGMT_LOAD function for both plain RSA and RSA-PSS
keys now also check that the key to be loaded is the correct type,
and refuse to load it if it's not.

levitte committed:
squash! PROV: Refactor DER->key decoder

Furthermore, there are cases where the decoder might not get the key
type it expected.  This may happen when certain key types that share
the same OpenSSL structure may be mixed up somehow.  The known cases
are EC vs SM2 and RSA vs RSA-PSS.

To remedy, we add the possibility to specify a checking function that
can check if the key that was decoded meets decoder expectations.

levitte commented:
A number of fixups added.  @mattcaswell, would you mind trying again?

levitte committed:
fixup! PROV: Refactor DER->key decoder

slontis commented on crypto/x509/x_pubkey.c:
```diff
@@ -33,6 +34,14 @@ struct X509_pubkey_st {
+
+    /*
+     * d2i_X509_PUBKEY() does double parsing, first with OSSL_DECODER,
+     * and the with ASN1_item_d2i().  It leaves the result of OSSL_DECODER
```
then

levitte replied:
Fixed

mattcaswell commented:
> A number of fixups added. @mattcaswell, would you mind trying again?

Same problem.

levitte committed:
fixup! X509: Refactor X509_PUBKEY processing to include provider side keys

levitte commented:
But hold on, @mattcaswell...  is there point that the EVP_PKEY *must* be provider backed?  I thought the primary problem was that it now *may* be provider backed.  In other words, provider backed EVP_PKEYs were *completely inaccessible* when decoding an X509_PUBKEY before this change, i.e. other algos than our built in ones were impossible.

t8m commented:
> But hold on, @mattcaswell... is there point that the EVP_PKEY _must_ be provider backed? I thought the primary problem was that it now _may_ be provider backed. In other words, provider backed EVP_PKEYs were _completely inaccessible_ when decoding an X509_PUBKEY before this change, i.e. other algos than our built in ones were impossible.

If it is not provider backed, will it be automatically imported to provider if you do for example signature verification of a certificate with the X509_PUBKEY? IMO if that's the case, I think we can keep that.

mattcaswell commented:
It remains unclear to me why we would do things differently for key types that happen to have a legacy implementation vs for those that don't. If we've got the codepaths to handle provider based keys - why wouldn't we always do that if there is a provider available to do it?

levitte commented:
> If it is not provider backed, will it be automatically imported to provider if you do for example signature verification of a certificate with the X509_PUBKEY? IMO if that's the case, I think we can keep that.

Yes, this was implemented early.  The keys are exported to the provider that is to do the signature work if that provider doesn't match the key's provider (or that the key is legacy).  This is among the earliest things we implemented.

levitte commented:
> It remains unclear to me why we would do things differently for key types that happen to have a legacy implementation vs for those that don't. If we've got the codepaths to handle provider based keys - why wouldn't we always do that if there is a provider available to do it?

Generally speaking, priorities. Do operations work with legacy EVP_PKEYs as well as with provider native EVP_PKEYs?  In that case, it shouldn't matter whether X509_get_pubkey() one or the other.
The main problem here was that if the algorithm given by the AlgorithmID *isn't known by OpenSSL*, then `X509_get_pubkey()` would obviously return NULL...  was it not?

mattcaswell commented:
> Generally speaking, priorities (we've all enough on our plates as it is). Do operations work with legacy EVP_PKEYs as well as with provider native EVP_PKEYs? In that case, it shouldn't matter whether X509_get_pubkey() one or the other.
The main problem here was that if the algorithm given by the AlgorithmID isn't known by OpenSSL, then X509_get_pubkey() would obviously return NULL... was it not?

That remains the primary issue - but I don't understand why it is any harder to not just do the same in all cases. Surely this is easier? Why does it behave differently in these cases?

levitte commented:
> That remains the primary issue - but I don't understand why it is any harder to not just do the same in all cases. Surely this is easier? Why does it behave differently in these cases?

I'm frankly not sure what you mean with "different"

levitte commented:
I've tested the code that @mattcaswell used, and it turns out that the solution here isn't right.  I'll have to revisit this

levitte commented:
> I'll have to revisit this

I knew there had to be a way:

https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L63-L66
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L609-L610
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L652-L660

Using that form allows a programmatic "take-over" of the processing of an item.  X509_NAME is processed this way, and it seems like the solution for our problem here is to do the same for X509_PUBKEY...

slontis commented on providers/implementations/encode_decode/decode_der2key.c:
```diff
@@ -277,59 +288,66 @@ static int der2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
-            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        goto next;
+
+    /* We try the typs specific functions first, if available */
```
type

slontis commented on crypto/x509/x_pubkey.c:
```diff
@@ -91,9 +105,52 @@ ASN1_SEQUENCE_cb(X509_PUBKEY, pubkey_cb) = {
+    if ((dctx = OSSL_DECODER_CTX_new_by_EVP_PKEY(&pkey,
+                                                 "DER", "SubjectPublicKeyInfo",
+                                                 NULL, EVP_PKEY_PUBLIC_KEY,
+                                                 NULL, NULL)) != NULL
```
Why is there no libctx, propq here?

slontis commented on providers/implementations/encode_decode/decode_der2key.c:
```diff
@@ -456,8 +489,18 @@ static void dsa_adjust(void *key, struct der2key_ctx_st *ctx)
 
+static int ec_check(void *key, struct der2key_ctx_st *ctx)
+{
+    /* We're trying to be clever by comparing two truths */
```
This comment seems to be a riddle..

levitte replied:
Not much, read the code...

levitte commented:
I've made this WIP because of the needed rework

levitte commented:
I'm closing in favor of a new PR where I've started over

levitte commented:
See #13281

slontis commented:
> See #13281

You mean  #14281

levitte commented:
Ah, oops, yes
